toolbox/timecomparer/main.go

```diff
@@ -0,0 +1,537 @@
+func readSpreadSheet(ctx con.Context, apiKey string, spreadsheetID string, readRange string) ([]string, error) {
+	srv, err := sheets.NewService(ctx, option.WithAPIKey(apiKey))
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve Sheets client: %v", err)
+	}
+
+	resp, err := srv.Spreadsheets.Values.Get(spreadsheetID, readRange).Do()
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve data from sheet: %v", err)
+	}
+
+	if len(resp.Values) == 0 {
+		return nil, fmt.Errorf("no data found")
+	}
+	var filePaths []string
+
+	for _, row := range resp.Values {
+		filePath := row[0]
+		filePathString, ok := filePath.(string)
+		if ok {
+			filePaths = append(filePaths, filePathString)
+		} else {
+			return nil, fmt.Errorf("file path is not string")
+		}
+	}
+
+	return filePaths, nil
+}
+
+func NewTimeComparer(client *storage.Client, bucketName string) *TimeComparer {
+	return &TimeComparer{
+		client:     client,
+		bucketName: bucketName,
+	}
+}
+
+// Read from gcs the build-log.txt files into slice of lines.
+func (tc *TimeComparer) query(ctx context.Context, prefix string) ([]string, error) {
+	client := tc.client
+	bucket := client.Bucket(tc.bucketName)
+	buildFile := bucket.Object(prefix + "build-log.txt")
+
+	rc, err := buildFile.NewReader(ctx)
+	if err != nil {
+		return []string{}, err
+	}
+	defer rc.Close()
+	var lines []string
+
+	scanner := bufio.NewScanner(rc)
+	scanner.Split(bufio.ScanLines)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if len(line) == 0 {
+			continue
+		}
+		lines = append(lines, line)
+	}
+
+	return lines, nil
+}
+
+type runPair struct {
+	name       string
+	stringTime string
+	time       []float64
+}
+
+type runTime struct {
+	runTime []runPair
+	runPath string
+}
+
+func (r runTime) toStringSlice() []string {
+	rSlice := []string{r.runPath}
+	for _, rTime := range r.runTime {
+		rSlice = append(rSlice, rTime.name, rTime.stringTime)
+	}
+	return rSlice
+}
+
+type runCombination struct {
+	totalTime float64
+	runTimes  []runTime
+}
+
+type commandAndTime struct {
+	command    string
+	runStrings []string
+}
+
+// Write error map and warning map to csv file with given file path.
+func writeCSV(sortedArray []float64, readResult map[string][]string, timeMap map[float64]commandAndTime, fileName string) error {
+	var file *os.File
+	var err error
+	// If file already exists in the path, write to original file.
+	if _, err = os.Stat(fileName); err == nil {
+		file, err = os.OpenFile(fileName, os.O_RDWR, 0644)
+		if err != nil {
+			return fmt.Errorf("cannot open file %v", err)
+		}
+	} else {
+		file, err = os.Create(fileName)
+		if err != nil {
+			return fmt.Errorf("cannot open file %v", err)
+		}
+	}
+
+	defer file.Close()
+
+	writer := csv.NewWriter(file)
+	defer writer.Flush()
+
+	for _, sortedTime := range sortedArray {
+		comAndTime := timeMap[sortedTime]
+		emptyCom := commandAndTime{}
+		if reflect.DeepEqual(comAndTime, emptyCom) {
+			continue
+		}
+
+		com := comAndTime.command
```

Comment:
Seems redundant to have both comAndTime and com. Consider getting rid of comAndTime